File: 34.find-first-and-last-position-of-element-in-sorted-array.java
/*
 * @lc app=leetcode id=34 lang=java
 *
 * [34] Find First and Last Position of Element in Sorted Array
 */

// @lc code=start
class Solution {
    public int[] searchRange(int[] nums, int target) {
        if (nums.length == 0) {
            return new int[] { -1, -1 };
        }
        int left = 0, right = nums.length;
        int mid = left + ((right - left) >> 1);
        while (left < right) {
            if (target < nums[mid]) {
                right = mid;
            } else if (target > nums[mid]) {
                left = mid + 1;
            } else {
                break;
            }
            mid = left + ((right - left) >> 1);
        }
        if (mid >= nums.length) {
            return new int[] { -1, -1 };
        }
        if (nums[mid] != target) {
            return new int[] { -1, -1 };
        }
        left = right = mid;
        while (left > 0 && nums[left - 1] == target) {
            left -= 1;
        }
        while (right < nums.length - 1 && nums[right + 1] == target) {
            right += 1;
        }
        return new int[] { left, right };
    }
}
// @lc code=end

File: 34.find-first-and-last-position-of-element-in-sorted-array.cpp
/*
 * @lc app=leetcode id=34 lang=cpp
 *
 * [34] Find First and Last Position of Element in Sorted Array
 */

// @lc code=start

#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        if (nums.size() == 0) {
            return {-1, -1};
        }
        if (target < nums.front() || target > nums.back()) {
            return {-1, -1};
        }
        auto left = lower_bound(nums.begin(), nums.end(), target);
        auto right = upper_bound(nums.begin(), nums.end(), target) - 1;
        if (*left == target) {
            return {static_cast<int>(distance(nums.begin(), left)), static_cast<int>(distance(nums.begin(), right)) };
        } else {
            return {-1, -1};
        }

    }
};
// @lc code=end
